src/Orleans.Http/GrainIdFromRouteGrainProvider.cs

```diff
@@ -47,27 +47,29 @@ public Task<IGrain> GetGrain(Type grainType)
                     {
                         case GrainIdType.String:
                             string stringId = (string)grainIdParameter;
-                            return Task.FromResult(this._clusterClient.GetGrain(grainType, stringId));
+                            return new ValueTask<IGrain>(this._clusterClient.GetGrain(grainType, stringId));
                         case GrainIdType.Integer:
                             long integerId = Convert.ToInt64(grainIdParameter);
-                            return Task.FromResult(this._clusterClient.GetGrain(grainType, integerId));
+                            return new ValueTask<IGrain>(this._clusterClient.GetGrain(grainType, integerId));
                         case GrainIdType.IntegerCompound:
-                            return Task.FromResult(this._clusterClient.GetGrain(grainType, Convert.ToInt64(grainIdParameter), (string)grainIdExtensionParameter));
+                            return new ValueTask<IGrain>(this._clusterClient.GetGrain(grainType, Convert.ToInt64(grainIdParameter), (string)grainIdExtensionParameter));
                         case GrainIdType.GuidCompound:
-                            return Task.FromResult(this._clusterClient.GetGrain(grainType, Guid.Parse((string)grainIdParameter), (string)grainIdExtensionParameter));
+                            return new ValueTask<IGrain>(this._clusterClient.GetGrain(grainType, Guid.Parse((string)grainIdParameter), (string)grainIdExtensionParameter));
                         default:
-                            return Task.FromResult(this._clusterClient.GetGrain(grainType, Guid.Parse((string)grainIdParameter)));
+                            return new ValueTask<IGrain>(this._clusterClient.GetGrain(grainType, Guid.Parse((string)grainIdParameter)));
                     }
                 }
 
                 context.Response.StatusCode = (int)System.Net.HttpStatusCode.BadRequest;
-                return Task.FromResult<IGrain>(null);
+                IGrain nullResult = null;
+                return new ValueTask<IGrain>(nullResult);
             }
             catch (Exception exc)
             {
                 context.Response.StatusCode = (int)System.Net.HttpStatusCode.BadRequest;
                 this._logger.LogError(exc, $"Failure getting grain '{grainType.FullName}' for route '{pattern.RawText}': {exc.Message}");
-                return Task.FromResult<IGrain>(null);
+                IGrain nullResult = null;
```

Comment:
This could be made a `private static readonly ValueTask<IGrain> _nullGrainReference = default; ` field and referenced everywhere is needed.

Reply:
Made the request change 👍